### PR TITLE
Fix k8s reconnect

### PR DESCRIPTION
**Pull request by djschaap**
With existing behavior, when a Cassandra node changes its IP address (while keeping the same host ID), the gocql client never utilizes that new IP address.  In the event ALL nodes change their IP addresses (over the lifetime of the gocql client), the cluster connection will become unusable.  Attempts to query Cassandra will be met with `gocql: no hosts available in the pool` errors.

I am experiencing this behavior connecting to a Cassandra cluster running on Kubernetes.  When a rolling restart occurs (StatefulSet change or otherwise), many of my long-lived gocql clients will become unresponsive.  Restarting the affected processes is required to restore Cassandra connectivity.

With these changes in place, when a `NEW_NODE` message is received for an existing host (host ID) with a new IP address, 1) the existing host (`HostInfo` struct) is removed and 2) the new host (`HostInfo` struct) is added.

Personal testing has shown this fix to work (i.e., the client learns the each new node IP address and maintains connectivity to the Cassandra cluster); more thorough testing would be welcome.

This may address #915 and/or #1582.

### CODE REVIEW

**Comment by martin-sucha** [comment]
Hello! Thank you for the pull request.

You have a very good point that `refreshRing` needs to handle re-adding the hosts, so I'll merge https://github.com/gocql/gocql/pull/1682/commits/8593ee86bc9f5fd9523c4d74637702d0e7621d17

As for the handling of events, please see #1680, what do you think about merging that pull request instead?

**Comment by lwimmer** [comment]
Thank you!

I'be briefly tested this PR and it seems to fix the issue described in #915 for us.

I've also tested #1680 and #1669, which did not fix the issue.

**Comment by djschaap** [comment]
@martin-sucha At first glance, I'm not thrilled with refreshing the ring on every topology change.  I share your fears re: thrashing.  It also feels bad/inefficient to disregard the actual information in a NEW_NODE message (i.e., knowledge of the host that is actually added) and trigger a full-scale refresh.

OTOH, doing a full ring refresh might help with consistency when the client misses a node add or remove event (if that ever occurs - I can't say).

Just to ask - would either approach interfere or fail when `DisableInitialHostLookup` or `DisableTopologyEvents` is true?  Are there any race/locking conditions that would be more likely to be triggered by doing a full ring refresh (vs a simple remove/add node)?

TL;DR I prefer my targeted approach to handling topology events, but #1680 looks like it would be sufficient for the scenarios I've considered.  I view both as functionally equivalent.

**Comment by Bolodya1997** [comment]
Hi, in our project we are extremely interested in this fix.

Is there any plan for merging this and adding to the new release? If so, when approximately it may become available?

**Comment by martin-sucha** [comment]
All right, I've merged this PR together with #1669 so that we can cut a bug fix release. It seems the change to update the existing host's IP is needed to fix #915. The arguments for refreshing the whole ring (with the debouncer) are compelling though, since it will reduce the amount of code that needs to be maintained. @joao-r-reis could you please rebase #1680 and add the debouncer? I think with 8593ee86bc9f5fd9523c4d74637702d0e7621d17 included #1680 should work as a fix to #915 too. We can release those changes in a subsequent release, that way even if someone encounters an issue with refreshing the ring always, they could fall back to a release that contains only #1682 and #1669.

**Comment by martin-sucha** [comment]
Tagged [1.3.2](https://github.com/gocql/gocql/releases/tag/v1.3.2)
